src/Raven.Server/Documents/Indexes/IndexStorage.cs

```diff
@@ -314,6 +329,95 @@ public static IndexType ReadIndexType(int indexId, StorageEnvironment environmen
             }
         }
 
+        public IEnumerable<string> GetReferencedCollections(RavenTransaction tx)
+        {
+            var tree = tx.InnerTransaction.ReadTree(Schema.ReferencedCollections);
+            using (var it = tree.Iterate(prefetch: true))
+            {
+                if (it.Seek(Slices.BeforeAllKeys) == false)
+                    yield break;
+
+                do
+                {
+                    yield return it.CurrentKey.ToString(); // TODO [ppekrol] ?
+                } while (it.MoveNext());
+            }
+        }
+
+        public IEnumerable<string> GetDocumentKeysThatReference(LazyStringValue key, RavenTransaction tx)
+        {
+            var reverseTree = tx.InnerTransaction.ReadTree(Schema.ReverseReferences);
+            using (var it = reverseTree.MultiRead(key))
+            {
+                if (it.Seek(Slices.BeforeAllKeys) == false)
+                    yield break;
+
+                do
+                {
+                    yield return it.CurrentKey.ToString(); // TODO [ppekrol] ?
+                } while (it.MoveNext());
+            }
+        }
+
+        public long ReadLastSeenEtagForReference(string collection, LazyStringValue key, RavenTransaction tx)
```

Comment:
Why have a separate key for collection here? Why not a LastSeenEtag tree for all collections?